keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,1060 @@
+container images.
+
+Some runtimes, like cri-o, mitigate these problems by using the `metacopy`
+option of overlayfs. This option avoids copying the whole file content when an
+operation updating only the metadata, like `chown` or `chmod`, is performed.
+This solution could be adopted by other runtimes until a more sophisticated
+approach is implemented in the kernel. [We
+talked](https://lists.linuxfoundation.org/pipermail/containers/2020-September/042230.html)
+to the Linux kernel community and [they
+replied](https://lists.linuxfoundation.org/pipermail/containers/2020-September/042230.html)
+they are working on a solution for it. If the Linux kernel provides a solution
+for this problem, that would be something that container runtimes should use. It
+does not impact the kubelet nor the CRI gRPC spec.
+
+#### Container Images with High IDs
+
+There are container images designed to run with high user and group IDs. It's
+possible that the IDs range assigned to the pod is not big enough to accommodate
+these IDs, in this case they will be mapped to the `nobody` user in the host.
+
+It's not a big problem in the `Cluster` case, the users have to be sure that
+they provide a range accommodating these IDs. It's more difficult to handle in
+the `Pod` case as the logic to allocate the ranges for each pod has to take this
+information in consideration. It's likely that this requires some changes to the
+CRI and kubelet so the runtimes can inform the kubelet what are the IDs present
+on a specific container image.
+
+## Implementation Phases
+
+The implementation of this KEP in a single phase is complicated as there are many
+discussions to be done. We learned from previous attempts to bring this support in
+that it should be done in small steps to avoid losing the focus on the
+discussion. It's also true that a full plan should be agreed at the beginning to
+avoid changing the implementation drastically in further phases.
+
+This proposal implementation aims to be divided in the following phases:
+
+### Phase 1
+
+This first phase includes:
+ - Extend the PodSpec with the `userNamespaceMode` field.
+ - Extend the CRI with user and ID mappings fields.
+ - Implement support for `Host` and `Cluster` user namespace modes.
+
+The goal of this phase is to implement some initial user namespace support
+providing pod-to-host isolation and supporting volumes. The implementation of
+the `Pod` mode is out of scope in this phase because it requires a non
+negligible amount of work and we could risk losing the focus failing to deliver
+this feature.
+
+### Future Phases
+
+These phases aim to implement the `Pod` mode. After these phases are completed the
+full advantanges of user namespaces can be used in some cases (stateless
+workloads).
+
+There are some things that have to be studied with more detail for these
+phase(s) but are not needed for phase 1, hence they are not discussed in detail:
+
+- **Pod Default Mode**:
```

Comment:
We should add a section on metacopyup here.

Reply:
Done in https://github.com/kinvolk/kubernetes-enhancements/pull/3/commits/bd9d2efbcf2580c404f717c02821d258c6d98c52. Please let me know if that's ok.